soh/src/code/z_play.c

```diff
@@ -431,6 +431,7 @@ void Play_Init(GameState* thisx) {
     // Invalid entrance, so immediately exit the game to opening title
     if (gSaveContext.entranceIndex == -1) {
         gSaveContext.entranceIndex = 0;
+        SaveManager_ThreadPoolWait();
```

Comment:
Might want to add this to `game.c` instead so that we synchronize saves whenever the game state is destructed. This happens on scene transitions. Any thoughts? @briaguya-ai @Jack-Walker @leggettc18 @Rozelette

Reply:
I'm thinking that doing it on all scene transitions would work well with the asset lifetime v2 issue https://github.com/HarbourMasters/Shipwright/issues/2757

Reply:
@Kenix3 I'm not sure why we would want that to happen on scene transition, and I'm not sure how it would fit in with the asset lifetime v2 stuff, could you elaborate on your thoughts here?

Reply:
It fits in because we'd be synchronizing on an async task to hide that during the scene transition. That's it.

Reply:
> The question is whether we need it on non-exiting transitions. Despite the idea for doing asynchronous asset loading, I would tend to think that the file saving thread pool should be isolated from any other threaded functionality in SoH. Not to mention that the idea was to prevent save lag, and putting a wait specifically in `GameState_Destroy`, if it is called on every transition, would merely reintroduce that lag in a different way.

It would still be running asynchronously. This would just guarantee that by the time you transition to a new scene, any saves that are in progress are completed.

Reply:
But again, why would that be necessary for saves, specifically?

Reply:
Because the more often that we save the less data corruption is going to happen.

Reply:
Unrelated to the "should we save on scene transition" question, what's the point of calling `SaveManager_ThreadPoolWait();` when `gSaveContext.entranceIndex == -1` in `Play_Init`?

Reply:
So, to clarify, you're wanting to add an autosave into `GameState_Destroy` regardless of a person's autosave selection? Are you intending to remove the autosave option? Either one would be a huge mistake. Saving any section more often than what the autosave setting is doing should be up to the parts of the code that add that section's handlers.

Reply:
> So, to clarify, you're wanting to add an autosave into `GameState_Destroy` regardless of a person's autosave selection? 

calling `SaveManager_ThreadPoolWait();` doesn't autosave, it makes sure any async saves that are currently in progress finish. either way, we don't need to add it to `GameState_Destroy`

Reply:
> Unrelated to the "should we save on scene transition" question, what's the point of calling `SaveManager_ThreadPoolWait();` when `gSaveContext.entranceIndex == -1` in `Play_Init`?

That was a point where `GameInteractor::OnExitGame` was being called, I was just covering bases. I don't know that I've ever encountered that particular situation. It was just that leaving it to any `OnExitGame` process after the playstate reset had already happened was still causing crashes when a save was in progress as an exit happened, so I put that ahead of any of that happening wherever the hook was called.

Reply:
> > So, to clarify, you're wanting to add an autosave into `GameState_Destroy` regardless of a person's autosave selection?
> 
> calling `SaveManager_ThreadPoolWait();` doesn't autosave, it makes sure any async saves that are currently in progress finish. either way, we don't need to add it to `GameState_Destroy`

Which makes its inclusion there even more perplexing, to be honest. There's nothing about a transition that would require waiting for save threads to finish, and like I said, that would just potentially reintroduce the save lag this was meant to eliminate.

Reply:
yeah, let's not force sync on scene transition

Reply:
> yeah, let's not force sync on scene transition

Sure.

Reply:
> That was a point where GameInteractor::OnExitGame was being called, I was just covering bases. I don't know that I've ever encountered that particular situation. It was just that leaving it to any OnExitGame process after the playstate reset had already happened was still causing crashes when a save was in progress as an exit happened, so I put that ahead of any of that happening wherever the hook was called.

so registering a hook for `OnExitGame` like we do with `ClearEntranceTrackingData` here was crashing?
https://github.com/HarbourMasters/Shipwright/blob/3f896f126ff7f36dff8afc2459104ff1adab548b/soh/soh/Enhancements/randomizer/randomizer_entrance_tracker.cpp#L928-L930

Reply:
Yes. Though, to be fair, I had other problems I had to diagnose at the same time, so there is a possibility that problem was being addressed elsewhere. I'm at work and have been setting up my computer with Atlas OS to try to eke out some more performance, I will revert it when I get home to make sure it was actually necessary.